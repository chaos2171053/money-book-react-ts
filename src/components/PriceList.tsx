import React from 'react'
import Ionicon from 'react-ionicons'


interface IProps {
    items: Array<{ id: number, title: string, price: number, date: string, category: { id: number, name: string, type: string, iconName: string } }>,
    onModifyItem: (item: { id: number, title: string }) => void,
    onDeleteItem: (item: { id: number }) => void
}

const PriceList = (props: IProps) => {
    return (
        <ul className="list-group list-group-flush">
            {
                props.items.map((item) => (
                    <li className="list-group-item d-flex justify-content-between align-items-center" key={item.id}>
                        <span className="col-1">
                            <Ionicon
                                className="rounded-circle"
                                fontSize="30px"
                                style={{ backgroundColor: '#007bff', padding: '5px' }}
                                color={'#fff'}
                                icon={item.category.iconName}>
                                {item.category.name}
                            </Ionicon>
                        </span>
                        <span className="col-5"> {item.title}</span>
                        <span className="col-2 font-weight-bold">
                            {(item.category.type === 'income') ? '+' : '-'}
                            {item.price}å…ƒ
                        </span>
                        <span className="col-2">{item.date}</span>
                        <button className="col-1" onClick={() => props.onModifyItem(item)}>
                            <Ionicon
                                className="rounded-circle"
                                fontSize="30px"
                                style={{ backgroundColor: '#28a745', padding: '5px' }}
                                color={'#fff'}
                                icon='ios-create-outline'>

                            </Ionicon>
                        </button>
                        <button className="col-1 " onClick={() => props.onDeleteItem(item)} >
                            <Ionicon
                                className="rounded-circle"
                                fontSize="30px"
                                style={{ backgroundColor: '#dc3545', padding: '5px' }}
                                color={'#fff'}
                                icon='ios-close'
                            />
                        </button>
                    </li>
                ))
            }
        </ul>
    )
}
export default PriceList;
