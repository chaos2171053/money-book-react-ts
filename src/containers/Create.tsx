import React, { Component } from 'react'
import PriceForm from '../components/PriceForm'
import { testCategories } from '../testData'
import { Tabs, Tab } from '../components/Tabs'
import { TYPE_OUTCOME } from '../utility'
import CategorySelect from '../components/CategorySelect'
import WithContext from '../components/WithContext'
import { RouteComponentProps, withRouter } from 'react-router-dom'
interface Ipros {

}
class CreatePage extends Component<Ipros & RouteComponentProps>{
    render() {
        const filterCategories = testCategories.filter(category => {
            return category.type === TYPE_OUTCOME
        })
        return (
            <div className="create-page py-3 px-3 rounded mt-3" style={{ background: '#fff' }}>
                <Tabs activeIndex={0} onTabChange={() => { }}>
                    <Tab>支出</Tab>
                    <Tab>收入</Tab>
                </Tabs>
                <CategorySelect categories={filterCategories}
                    onSelectCategory={() => { }}
                />
                <PriceForm onFormSubmit={() => { }} onCancelSubmit={() => { }} ></PriceForm>
            </div >
        )
    }
}

export default WithContext(withRouter(CreatePage))