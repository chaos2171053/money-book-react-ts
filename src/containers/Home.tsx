import React, { Component } from 'react';
import { RouteComponentProps, withRouter } from 'react-router-dom'
import PriceList from '../components/PriceList'
import { LIST_VIEW, CHART_VIEW, TYPE_OUTCOME } from '../utility'
import MonthPicker from '../components/MonthPicker'
import TotalPrice from '../components/TotalPrice';
import CreateBtn from '../components/CreateBtn'
import { Tab, Tabs } from '../components/Tabs'
import Ionicon from 'react-ionicons'
import WithContext from '../components/WithContext';
import Loader from '../components/Loader'
interface Category {
    id: string,
    name: string,
    type: string,
    iconName: string
}

export interface PriceItem {
    id: string,
    title: string,
    price: number,
    date: string,
    cid: string,
    category: { id: string, name: string, type: string, iconName: string }
}



const tabsText = [LIST_VIEW, CHART_VIEW]

interface IProps extends RouteComponentProps {
    data: {
        items: {
            [key: string]: PriceItem
        },
        categories: {
            [key: string]: Category
        },
        currentDate: {
            year: number,
            month: number
        },
        isLoading: boolean
    },
    actions: {
        deleteItem: Function,
        getInitalData: Function,
        selectNewMonth: Function
    }
}

interface IState {
    tabView: string
}


class HomePage extends Component<IProps, IState> {
    static defaultProps = {
        data: {
            items: {},
            categories: {},
            currentDate: {
                year: 0,
                month: 0
            },
            isLoading: false
        }
    }
    constructor(props: IProps) {
        super(props)
        this.state = {
            // items,
            tabView: tabsText[0],
        }
    }
    componentDidMount() {
        this.props.actions.getInitalData()
    }

    changeView = (index: number) => {
        this.setState({
            tabView: tabsText[index],
        })
    }
    changeDate = (year: number, month: number) => {
        this.props.actions.selectNewMonth(year, month)
    }
    modifyItem = (modifidItem: { title: string, id: string }) => {
        this.props.history.push(`/edit/${modifidItem.id}`)
    }

    createItem = () => {
        this.props.history.push('/create')
    }

    deteteItem = (deletedItem: { id: string }) => {
        this.props.actions && this.props.actions.deleteItem(deletedItem)
    }

    render() {
        const { data } = this.props

        let { items, categories, currentDate, isLoading } = data
        const { tabView } = this.state
        const itemsWithCategory = Object.keys(items).map((id: string) => {
            items[id].category = categories[items[id].cid]
            return items[id]
        })
        let totalIncome = 0, totalOutcome = 0

        itemsWithCategory && itemsWithCategory.forEach(item => {
            if (item.category.type === TYPE_OUTCOME) {
                totalOutcome += item.price
            } else {
                totalIncome += item.price
            }
        })

        return (
            <React.Fragment>
                <div className="row">
                    <div className="col">
                        <MonthPicker year={currentDate.year} month={currentDate.month} onChange={this.changeDate}></MonthPicker>
                    </div>
                    <div className="col">
                        <TotalPrice
                            income={totalIncome}
                            outcome={totalOutcome}
                        />
                    </div>
                </div>
                <div className="content-area py-3 px-3">
                    {isLoading &&
                        <Loader />
                    }
                    {!isLoading &&
                        <React.Fragment>
                            <Tabs activeIndex={0} onTabChange={this.changeView}>
                                <Tab>
                                    <Ionicon
                                        className="rounded-circle mr-2"
                                        fontSize="25px"
                                        color={'#007bff'}
                                        icon='ios-paper'
                                    />
                                    列表模式
                        </Tab>
                                <Tab>
                                    <Ionicon
                                        className="rounded-circle mr-2"
                                        fontSize="25px"
                                        color={'#007bff'}
                                        icon='ios-pie'
                                    />
                                    图表模式
                        </Tab>
                            </Tabs>
                            <CreateBtn onClick={this.createItem}></CreateBtn>
                            {
                                itemsWithCategory && tabView === LIST_VIEW && <PriceList items={itemsWithCategory} onModifyItem={this.modifyItem} onDeleteItem={this.deteteItem} />

                            }
                            {tabView === LIST_VIEW && itemsWithCategory.length === 0 &&
                                <div className="alert alert-light text-center no-record">
                                    您还没有任何记账记录
                    </div>
                            }
                            {
                                tabView === CHART_VIEW &&
                                <h1>这里是图表</h1>
                            }

                        </React.Fragment>
                    }
                </div>
            </React.Fragment>

        )
    }
}

export const Home = WithContext(withRouter(HomePage))



