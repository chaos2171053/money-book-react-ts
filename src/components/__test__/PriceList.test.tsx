import React from 'react'
import { shallow } from 'enzyme'
import PriceList from '../PriceList'
import { items, categoies } from '../../containers/Home'
import Ionicon from 'react-ionicons'


const itemsWithCategory = items.map((item) => {
    item.category = categoies[item.cid]
    return item
})

const props = {
    items: itemsWithCategory,
    onModifyItem: jest.fn(),
    onDeleteItem: jest.fn()
}

let wrapper: any

describe('test PriceList Compoent', () => {
    beforeEach(() => {
        wrapper = shallow(<PriceList {...props} />)
    })
    it('should render the component to match snapshot', () => {
        expect(wrapper).toMatchSnapshot()
    })
    it('should render correct price items length', () => {
        expect(wrapper.find('.list-group-item').length).toEqual(itemsWithCategory.length)
    })
    it('should render correct icon and price for each item', () => {
        const iconList = wrapper.find('.list-group-item').first().find(Ionicon)
        expect(iconList.length).toEqual(3)
        expect(iconList.first().props().icon).toEqual(itemsWithCategory[0].category.iconName)
    })
    it('should trigger the correct function callbacks', () => {
        const firstItem = wrapper.find('.list-group-item').first()
        firstItem.find('button').first().simulate('click', { preventDefault: () => { } })
        expect(props.onModifyItem).toHaveBeenCalledWith(itemsWithCategory[0])
        firstItem.find('button').last().simulate('click', { preventDefault: () => { } })
        expect(props.onDeleteItem).toHaveBeenCalledWith(itemsWithCategory[0])
    })
})
