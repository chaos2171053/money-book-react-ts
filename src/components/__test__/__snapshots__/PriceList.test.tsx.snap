// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test PriceList Compoent should render the component to match snapshot 1`] = `ShallowWrapper {}`;
